src/particles/diagnostics/CovarianceMatrixMath.H

```diff
@@ -0,0 +1,209 @@
+/* Copyright 2022-2023 The Regents of the University of California, through Lawrence
+ *           Berkeley National Laboratory (subject to receipt of any required
+ *           approvals from the U.S. Dept. of Energy). All rights reserved.
+ *
+ * This file is part of ImpactX.
+ *
+ * Authors: Chad Mitchell, Axel Huebl
+ * License: BSD-3-Clause-LBNL
+ */
+#ifndef COVARIANCE_MATRIX_MATH_H
+#define COVARIANCE_MATRIX_MATH_H
+
+#include <ablastr/constant.H>
+
+#include <AMReX_Array.H>
+#include <AMReX_REAL.H>
+#include <AMReX_GpuComplex.H>
+
+#include <cmath>
+#include <tuple>
+
+
+namespace impactx::diagnostics
+{
+
+    /** Function to return the roots of a cubic polynomial ax^3 + bx^2 + cx + d.
+     *  The trigonometric form of Cardano's formula is used.
+     *  This implementation expects three real roots, which is verified
+     *  by checking the sign of the discriminant.
+     *
+     * @param[in] a coefficient of cubic term
+     * @param[in] b coefficient of quadratic term
+     * @param[in] c coefficient of linear term
+     * @param[in] d coefficient of constant term
+     * @returns tuple of three real roots
+     */
+    std::tuple<
+            amrex::ParticleReal,
+            amrex::ParticleReal,
+            amrex::ParticleReal>
+    CubicRootsTrig (
+        amrex::ParticleReal a,
+        amrex::ParticleReal b,
+        amrex::ParticleReal c,
+        amrex::ParticleReal d
+    )
+    {
+        using namespace amrex::literals;
+        using ablastr::constant::math::pi;
+
+        std::tuple <amrex::ParticleReal,amrex::ParticleReal,amrex::ParticleReal> roots;
+        amrex::ParticleReal x1 = 0.0_prt;
+        amrex::ParticleReal x2 = 0.0_prt;
+        amrex::ParticleReal x3 = 0.0_prt;
+
+        amrex::ParticleReal Q = (3.0_prt*a*c - std::pow(b,2))/(9.0_prt * std::pow(a,2));
+        amrex::ParticleReal R = (9.0_prt*a*b*c - 27_prt*std::pow(a,2)*d - 2.0_prt*std::pow(b,3))/(54.0_prt*std::pow(a,3));
+        amrex::ParticleReal discriminant = std::pow(Q,3) + std::pow(R,2);
+
+        // Discriminant should be < 0.  Otherwise, keep theta at default and throw an error.
+        amrex::ParticleReal tol = 1.0e-12;  //allow for roundoff error
+        if(discriminant > tol){
+
+           std::cout << "Polynomial in CubicRoots has one or more complex roots." << "\n";
```

Comment:
This would fit well for the error channel:
```suggestion
           std::cerr << "impactx::diagnostics::CubicRootsTrig Polynomial has one or more complex roots." << "\n";
```

Reply:
Actually, even better, is to use our warning manager:
- include: https://github.com/ECP-WarpX/impactx/blob/4a7da60f72bb5149dca9d23924ee16dca7421d46/src/initialization/InitMeshRefinement.cpp#L15C32-L15C43
- usage:
https://github.com/ECP-WarpX/impactx/blob/4a7da60f72bb5149dca9d23924ee16dca7421d46/src/initialization/InitMeshRefinement.cpp#L51-L59

This one handles MPI well and groups warnings into user-friendly sections of the output.

Reply:
Ok, done.